src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs

```diff
@@ -787,31 +637,198 @@ public class ClassWithPropertiesAndFieldsThatAreReservedKeywords
         [Fact]
         public static void NoErrorsWhenUsingIgnoredReservedCSharpKeywords()
         {
-            string source = @"using System;
-using System.Text.Json.Serialization;
+            string source = """
+                using System;
+                using System.Text.Json.Serialization;
 
-namespace Test
-{
-    [JsonSerializable(typeof(ClassWithPropertyNameThatIsAReservedKeyword))]
-    public partial class JsonContext : JsonSerializerContext { }
+                namespace Test
+                {
+                    [JsonSerializable(typeof(ClassWithPropertyNameThatIsAReservedKeyword))]
+                    public partial class JsonContext : JsonSerializerContext { }
 
-    public class ClassWithPropertyNameThatIsAReservedKeyword
-    {
-        [JsonIgnore]
-        public string @event { get; set; }
-    }
-}
-";
-            Compilation compilation = CompilationHelper.CreateCompilation(source);
-            JsonSourceGenerator generator = new JsonSourceGenerator();
+                    public class ClassWithPropertyNameThatIsAReservedKeyword
+                    {
+                        [JsonIgnore]
+                        public string @event { get; set; }
+                    }
+                }
+                """;
 
-            Compilation newCompilation = CompilationHelper.RunGenerators(compilation, out _, generator);
-            ImmutableArray<Diagnostic> generatorDiags = newCompilation.GetDiagnostics();
+            Compilation compilation = CompilationHelper.CreateCompilation(source);
+            JsonSourceGeneratorResult result = CompilationHelper.RunJsonSourceGenerator(compilation);
+            ImmutableArray<Diagnostic> generatorDiags = result.NewCompilation.GetDiagnostics();
 
             // No diagnostics expected.
             CompilationHelper.CheckDiagnosticMessages(DiagnosticSeverity.Info, generatorDiags, Array.Empty<(Location, string)>());
             CompilationHelper.CheckDiagnosticMessages(DiagnosticSeverity.Warning, generatorDiags, Array.Empty<(Location, string)>());
             CompilationHelper.CheckDiagnosticMessages(DiagnosticSeverity.Error, generatorDiags, Array.Empty<(Location, string)>());
         }
+
+        [Theory]
+        [MemberData(nameof(GetCompilationHelperFactories))]
```

Comment:
New tests validating cacheability properties of the model types.

Reply:
There's also Roslyn APIs that explicitly validate that a given run of the generator is incremental with respect to the previous run. I think it's worthwhile to add tests using that API as well (and you can build these tests to run on that API instead of building something from scratch if you would like).

Reply:
Interesting, could you point me to examples?

Reply:
Here's the examples for the LibraryImportGenerator: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/IncrementalGenerationTests.cs

Here's some examples in Roslyn: https://github.com/dotnet/roslyn/blob/236149d501afb0e898069243a6418435e9140ee0/src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs#L1665

Basically you can name a step and then traverse the tree of steps to see the inputs to a step and its outputs and whether or not they were new/modified/unchanged/pulled from the cache/removed/etc.